File: Artificial Intelligence/c++/BFS.cpp
#include <iostream>
#include <queue>
#include <vector>
using namespace std;

int main() {
    int n = 6;

    vector<vector<int>> graph = {
        {1, 2},
        {0, 3, 4},
        {0, 4},
        {1, 5},
        {1, 2, 5},
        {3, 4}
    };

    vector<bool> visited(n, false);
    queue<int> q;

    q.push(0);
    visited[0] = true;

    cout << "BFS Traversal: ";

    while (!q.empty()) {
        int node = q.front();
        q.pop();

        cout << node << " ";

        for (int neighbor : graph[node]) {
            if (!visited[neighbor]) {
                visited[neighbor] = true;
                q.push(neighbor);
            }
        }
    }

    return 0;
}

File: AI/Ex-Qn/G-I Solve/bfs.py
from collections import deque

graph = {
    'A': ['B', 'C', 'D'],
    'B': ['E', 'F'],
    'C': ['F'],
    'D': ['G'],
    'E': ['H'],
    'F': [],
    'G': [],
    'H': []
}

def bfs(graph, start, goal):
    visited = set()
    queue = deque([[start]])

    print("BFS Traversal Order:")
    while queue:
        path = queue.popleft()
        node = path[-1]

        if node not in visited:
            print(node, end=' ')
            visited.add(node)

            if node == goal:
                print(f"\n\nGoal {goal} found!")
                print(f"Path: {' → '.join(path)}")
                return path

            for neighbor in graph[node]:
                if neighbor not in visited:
                    new_path = list(path)
                    new_path.append(neighbor)
                    queue.append(new_path)

    print(f"\nGoal {goal} not reachable")
    return None

bfs(graph, 'A', 'F')
